openshift/action.go

```diff
@@ -0,0 +1,350 @@
+	return c.method
+}
+
+func (c *commonNamespaceAction) Sort(toSort environment.ByKind) {
+	sort.Sort(toSort)
+}
+
+func (c *commonNamespaceAction) Filter() FilterFunc {
+	return func(objects environment.Object) bool {
+		return true
+	}
+}
+
+func (c *commonNamespaceAction) ForceMasterTokenGlobally() bool {
+	return true
+}
+
+var NoHealing = func(openShiftService *ServiceBuilder) Healing {
+	return func(originalError error) error {
+		return originalError
+	}
+}
+
+func (c *commonNamespaceAction) HealingStrategy() HealingFuncGenerator {
+	return NoHealing
+}
+
+func (c *commonNamespaceAction) ManageAndUpdateResults(errorChan chan error, envTypes []environment.Type, healing Healing) error {
+	msg := utils.ListErrorsInMessage(errorChan)
+	if len(msg) > 0 {
+		err := fmt.Errorf("%s method applied to namespace types %s failed with one or more errors:%s", c.method, envTypes, msg)
+		if !c.allowSelfHealing {
+			return err
+		}
+		return healing(err)
+	}
+	return nil
+}
+
+func (c *Create) HealingStrategy() HealingFuncGenerator {
+	return func(openShiftService *ServiceBuilder) Healing {
+		return func(originalError error) error {
+			log.Error(openShiftService.service.context.requestCtx, map[string]interface{}{
+				"err":                   originalError,
+				"self-healing-strategy": "recreate-with-new-nsBaseName",
+			}, "the creation failed, starting self-healing logic")
+			openShiftUsername := openShiftService.service.context.openShiftUsername
+			tnnt, err := c.tenantRepo.GetTenant()
+			errMsgSuffix := fmt.Sprintf("while doing self-healing operations triggered by error: [%s]", originalError)
+			if err != nil {
+				return errors.Wrapf(err, "unable to get tenant %s", errMsgSuffix)
+			}
+			namespaces, err := c.tenantRepo.GetNamespaces()
+			fmt.Println(namespaces[0].ID)
+			if err != nil {
+				return errors.Wrapf(err, "unable to get namespaces of tenant %s %s", tnnt.ID, errMsgSuffix)
+			}
+			err = openShiftService.WithDeleteMethod(namespaces, true, true, true).ApplyAll(environment.DefaultEnvTypes)
+			if err != nil {
+				return errors.Wrapf(err, "deletion of namespaces failed %s", errMsgSuffix)
+			}
+			newNsBaseName, err := tenant.ConstructNsBaseName(c.tenantRepo.Service(), environment.RetrieveUserName(openShiftUsername))
+			if err != nil {
+				return errors.Wrapf(err, "unable to construct namespace base name for user with OSname %s %s", openShiftUsername, errMsgSuffix)
+			}
+			tnnt.NsBaseName = newNsBaseName
+			err = c.tenantRepo.Service().SaveTenant(tnnt)
+			if err != nil {
+				return errors.Wrapf(err, "unable to update tenant db entity %s", errMsgSuffix)
+			}
+			openShiftService.service.context.nsBaseName = newNsBaseName
+			err = openShiftService.WithPostMethod(false).ApplyAll(environment.DefaultEnvTypes)
+			if err != nil {
+				return errors.Wrapf(err, "unable to create new namespaces %s", errMsgSuffix)
+			}
+			return nil
+		}
+	}
+}
+
+func NewCreate(tenantRepo tenant.Repository, allowSelfHealing bool) *Create {
```

Comment:
Maybe rename `Create` to `CreateAction` ? `Create` alone looks a bit strange IMO

Reply:
renamed 822ba3b (the other action structs as well)